pecos/xmr/reranker/model.py

```diff
@@ -0,0 +1,420 @@
+
+    def __init__(self, config: CrossEncoderConfig):
+        """
+        Initialize the cross encoder model
+        Args:
+            hf_model: The huggingface model
+            train_group_size: The group size for training (number of documents per query)
+        """
+        super().__init__(config)
+        base_model = AutoModelForSequenceClassification.from_pretrained(
+            config.model_shortcut, num_labels=1, **config.model_init_kwargs
+        )
+        base_model.config.pad_token_id = (
+            0 if base_model.config.pad_token_id is None else base_model.config.pad_token_id
+        )
+        self.hf_model = base_model
+
+    @classmethod
+    def from_pretrained(
+        cls,
+        pretrained_model_name_or_path: Union[str, os.PathLike],
+        *model_args,
+        config: Optional[Union[PretrainedConfig, str, os.PathLike]] = None,
+        cache_dir: Optional[Union[str, os.PathLike]] = None,
+        ignore_mismatched_sizes: bool = False,
+        force_download: bool = False,
+        local_files_only: bool = False,
+        token: Optional[Union[str, bool]] = None,
+        revision: str = "main",
+        use_safetensors: Optional[bool] = None,
+        **kwargs,
+    ):
+        """
+        Load the model from the pretrained model name or path
+        """
+        is_local = os.path.isdir(pretrained_model_name_or_path)
+        param_folder = pretrained_model_name_or_path
+
+        def super_return():
+            return PreTrainedModel.from_pretrained(
+                pretrained_model_name_or_path,
+                *model_args,
+                config,
+                cache_dir,
+                ignore_mismatched_sizes,
+                force_download,
+                local_files_only,
+                token,
+                revision,
+                use_safetensors,
+                **kwargs,
+            )
+
+        if not is_local:
+            raise NotImplementedError(f"{cls} can only load local models")
+
+        with open(os.path.join(param_folder, PARAM_FILENAME), "r") as param_file:
+            params = json.load(param_file)
+
+        xe_config = CrossEncoder.Config.from_dict(params["model_params"]["encoder_args"])
+        xe_config = CrossEncoderConfig(**xe_config.to_dict())
+        model = CrossEncoder(xe_config)
+
+        try:
+            if xe_config.model_modifier["modifier_type"] == "peft":
+                model = PeftModel.from_pretrained(model, param_folder)
+            else:
+                super_return()
+        except KeyError:
+            logger.info("No peft configuration found")
+
+        return model
+
+    def forward(self, *args, **kwargs):
+        """
+        Returns the forward output of the huggingface model
+        """
+        return self.hf_model(*args, **kwargs)
+
+    def gradient_checkpointing_enable(self, **kwargs):
+        """
+        Enable gradient checkpointing for the model
+        """
+        self.hf_model.enable_input_require_grads()
```

Comment:
Not sure if this is the right place to call `hf_model.enable_input_require_grads()`.
From Tevatron RankLlaMA implementation (https://github.com/texttron/tevatron/blob/main/src/tevatron/reranker/modeling.py#L79), they are calling only when both "LoRA" and "training_args.gradient_checkpointing" is enable.

Reply:
The following sequence of operation is expected:
1. First get_modifed_model is called.
2. the trainer calls `gradient_checkpointing_enable`, when gradient checkpointing is enabled.